src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php

```diff
@@ -45,179 +42,163 @@ final class PhpStormStubsSourceStubber implements SourceStubber
             $stub = str_replace(' extends \iterable', '', $stub);
         }
 
-        return $stub;
+        return new StubData($stub, $this->getExtensionFromFilePath($filePath));
     }
 
     /**
      * {@inheritDoc}
      */
-    public function generateFunctionStub(CoreReflectionFunction $functionReflection) : ?string
+    public function generateFunctionStub(string $functionName) : ?StubData
     {
-        if ($functionReflection->isUserDefined()) {
+        if (! array_key_exists($functionName, $this->functionStubsMap)) {
             return null;
         }
 
-        return $this->getStub($functionReflection->getExtensionName(), $this->getFunctionNodeVisitor($functionReflection));
-    }
+        $filePath = $this->functionStubsMap[$functionName];
 
-    private function getStub(string $extensionName, NodeVisitorAbstract $nodeVisitor) : ?string
-    {
-        $node = null;
+        if (! array_key_exists($functionName, $this->functionNodes)) {
+            $this->parseFile($filePath);
+        }
 
-        $this->nodeTraverser->addVisitor($nodeVisitor);
+        return new StubData($this->createStub($this->functionNodes[$functionName]), $this->getExtensionFromFilePath($filePath));
+    }
 
-        foreach ($this->getExtensionStubsFiles($extensionName) as $filePath) {
-            FileChecker::assertReadableFile($filePath);
+    private function parseFile(string $filePath) : void
+    {
+        $ast = $this->phpParser->parse(file_get_contents($this->getAbsoluteFilePath($filePath)));
 
-            $ast = $this->phpParser->parse(file_get_contents($filePath));
+        /** @psalm-suppress UndefinedMethod */
+        $this->cachingVisitor->clearNodes();
 
-            $this->nodeTraverser->traverse($ast);
+        $this->nodeTraverser->traverse($ast);
 
-            /** @psalm-suppress UndefinedMethod */
-            $node = $nodeVisitor->getNode();
-            if ($node !== null) {
-                break;
-            }
+        /** @psalm-suppress UndefinedMethod */
+        foreach ($this->cachingVisitor->getClassNodes() as $className => $classNode) {
+            $this->classNodes[$className] = $classNode;
         }
 
-        $this->nodeTraverser->removeVisitor($nodeVisitor);
-
-        if ($node === null) {
-            return null;
+        /** @psalm-suppress UndefinedMethod */
+        foreach ($this->cachingVisitor->getFunctionNodes() as $functionName => $functionNode) {
+            $this->functionNodes[$functionName] = $functionNode;
         }
+    }
 
+    private function createStub(Node $node) : string
+    {
         return "<?php\n\n" . $this->prettyPrinter->prettyPrint([$node]) . "\n";
     }
 
-    private function getClassNodeVisitor(CoreReflectionClass $classReflection) : NodeVisitorAbstract
+    private function createCachingVisitor() : NodeVisitorAbstract
     {
-        return new class($classReflection->getName()) extends NodeVisitorAbstract
+        return new class() extends NodeVisitorAbstract
```

Comment:
One might say it's not BetterReflection but this piece is a deal-breaker currently for Box. Indeed since it is an anonymous class, this object cannot be serialized which is causing issues when leveraging parallelisation in Box. Would it be possible to move it to a dedicated class? If the API stability is a concern, it could be annotated with a `@internal` tag

Reply:
Uhhh, that is `private` state of this particular object: why is this breaking Box?